File: labs/mobile/donor-archive/tunnel/AmneziaObfsParameters.kt
package com.luminet.android.tunnel

data class AmneziaObfsConfig(
    var jc: Int = 4,
    var jmin: Int = 40,
    var jmax: Int = 70,
    var s1: Int = 56,
    var s2: Int = 56,
    var h1: Long = 0x01000000L,
    var h2: Long = 0x02000000L,
    var h3: Long = 0x03000000L,
    var h4: Long = 0x04000000L
) {
    fun isValid(): Boolean {
        if (jc !in 0..128) return false
        if (jmin < 0 || jmax < jmin || jmax > 1500) return false
        if (s1 !in 0..1024 || s2 !in 0..1024) return false
        return true
    }

    fun parseLine(line: String): Boolean {
        val parts = line.split("=")
        if (parts.size != 2) return false
        val key = parts[0].trim().uppercase()
        val rawVal = parts[1].trim()

        val num = try {
            if (rawVal.startsWith("0x", ignoreCase = true)) {
                rawVal.substring(2).toLong(16)
            } else {
                rawVal.toLong()
            }
        } catch (_: Exception) {
            return false
        }

        when (key) {
            "JC" -> jc = num.toInt()
            "JMIN" -> jmin = num.toInt()
            "JMAX" -> jmax = num.toInt()
            "S1" -> s1 = num.toInt()
            "S2" -> s2 = num.toInt()
            "H1" -> h1 = num
            "H2" -> h2 = num
            "H3" -> h3 = num
            "H4" -> h4 = num
            else -> return false
        }
        return true
    }
}

File: labs/control-ui/api-shadow/amnezia_obfs_parameters.ts
export interface AmneziaObfsConfig {
  jc: number;
  jmin: number;
  jmax: number;
  s1: number;
  s2: number;
  h1: number;
  h2: number;
  h3: number;
  h4: number;
}

export class AmneziaObfsParameters {
  public params: AmneziaObfsConfig = {
    jc: 4,
    jmin: 40,
    jmax: 70,
    s1: 56,
    s2: 56,
    h1: 0x01000000,
    h2: 0x02000000,
    h3: 0x03000000,
    h4: 0x04000000,
  };

  isValid(): boolean {
    if (this.params.jc < 0 || this.params.jc > 128) return false;
    if (this.params.jmin < 0 || this.params.jmax < this.params.jmin || this.params.jmax > 1500)
      return false;
    if (this.params.s1 < 0 || this.params.s1 > 1024 || this.params.s2 < 0 || this.params.s2 > 1024)
      return false;
    return true;
  }

  parseLine(line: string): boolean {
    const parts = line.split('=');
    if (parts.length !== 2) return false;

    const key = parts[0]!.trim().toUpperCase();
    const rawVal = parts[1]!.trim();

    let num: number;
    if (rawVal.toLowerCase().startsWith('0x')) {
      num = parseInt(rawVal.substring(2), 16);
    } else {
      num = parseInt(rawVal, 10);
    }
    if (isNaN(num)) return false;

    switch (key) {
      case 'JC':
        this.params.jc = num;
        break;
      case 'JMIN':
        this.params.jmin = num;
        break;
      case 'JMAX':
        this.params.jmax = num;
        break;
      case 'S1':
        this.params.s1 = num;
        break;
      case 'S2':
        this.params.s2 = num;
        break;
      case 'H1':
        this.params.h1 = num >>> 0;
        break;
      case 'H2':
        this.params.h2 = num >>> 0;
        break;
      case 'H3':
        this.params.h3 = num >>> 0;
        break;
      case 'H4':
        this.params.h4 = num >>> 0;
        break;
      default:
        return false;
    }
    return true;
  }
}
